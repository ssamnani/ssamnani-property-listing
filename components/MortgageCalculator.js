import React, { useState, useEffect } from "react";
import { Input } from "antd";
import MortgageDownPaymentMetricToggle from "./MortgageDownPaymentMetricToggle";
import "../styles/MortgageCalculator.module.css";

function MortgageCalculator() {
  let inputStyleWide = {
    width: 200,
    backgroundColor: "rgb(243, 244, 246)",
  };
  let inputStyleSmall = {
    width: 120,
    backgroundColor: "rgb(243, 244, 246)",
    minWidth: "100px",
  };

  // States
  const [monthlyAmount, setMonthlyAmount] = useState(1886);
  const [loanAmount, setLoanAmount] = useState(450000);
  const [propertyPrice, setPropertyPrice] = useState(500000);
  const [downPayment, setDownPayment] = useState(10);
  const [downPaymentMetric, setDownPaymentMetric] = useState("%");
  const [interestRate, setInterestRate] = useState(3.6);
  const [loanYears, setLoanYears] = useState(35);

  useEffect(() => {
    calculateMonthlyPayment();
  }, [propertyPrice, downPayment, downPaymentMetric, interestRate, loanYears]);

  const calcLoanAmount = () => {
    return downPaymentMetric === "%"
      ? propertyPrice - (propertyPrice * downPayment) / 100
      : propertyPrice - downPayment;
  };

  const calculateMonthlyPayment = () => {
    let loanAmount = calcLoanAmount();
    setLoanAmount(loanAmount);
    const loanMonths = loanYears * 12;
    let i = interestRate / 12 / 100;
    let monthly =
      (loanAmount * i * Math.pow(1 + i, loanMonths)) /
      (Math.pow(1 + i, loanMonths) - 1);
    setMonthlyAmount(monthly.toFixed(0));
  };

  const handlePropertyPriceChange = (event) => {
    setPropertyPrice(event.target.value);
    calculateMonthlyPayment();
  };

  const handleDownPaymentChange = (event) => {
    setDownPayment(event.target.value);
    calculateMonthlyPayment();
  };

  const handleInterestRateChange = (event) => {
    setInterestRate(event.target.value);
    calculateMonthlyPayment();
  };

  const handleLoanYearsChange = (event) => {
    setLoanYears(event.target.value);
    calculateMonthlyPayment();
  };

  const handleMetricChange = (val) => {
    if (val === downPaymentMetric) return;
    setDownPaymentMetric(val);
    val === "%"
      ? setDownPayment((downPayment / propertyPrice) * 100)
      : setDownPayment((downPayment * propertyPrice) / 100);
  };

  return (
    <div className="moCalculator">
      <div className="mortgageTitleBar">
        <span className="morgCalcTitle">Mortgage Calculator</span>
        <span className="companyLogoWrap">
          <img src="./assets/images/iproperty-logo-black.png" />
        </span>
      </div>
      <div className="moCalcBody">
        <div className="moCalcLeft">
          <div className="monthlyAmount">
            <span data-testid="monthly-amount">RM {monthlyAmount}</span>
            <span className="pm">p/m</span>
          </div>
          <div className="loanAmount">
            <span data-testid="loan-amount">Loan Amount: RM {loanAmount}</span>
          </div>
        </div>
        <div className="moCalcRight">
          <div className="moInputWrap">
            <div>
              <div className="moInpLabel">Property Price (RM)</div>
              <Input
                type="number"
                style={inputStyleWide}
                value={propertyPrice}
                onChange={handlePropertyPriceChange}
                data-testid="property-price"
              />
            </div>
            <div>
              <div className="moInpLabel">Down Payment</div>
              <div className="downPaymentInputsWrapper">
                <Input
                  type="number"
                  style={inputStyleSmall}
                  value={downPayment}
                  onChange={handleDownPaymentChange}
                  data-testid="down-payment"
                />
                <MortgageDownPaymentMetricToggle
                  metric={downPaymentMetric}
                  onChangeMetric={handleMetricChange}
                  data-testid="down-payment-metric"
                />
              </div>
            </div>
          </div>
          <div className="moInputWrap">
            <div>
              <div className="moInpLabel">Interest Rate (%)</div>
              <Input
                type="number"
                style={inputStyleWide}
                value={interestRate}
                onChange={handleInterestRateChange}
                data-testid="interest-rate"
              />
            </div>
            <div>
              <div className="moInpLabel">Loan term (Years)</div>
              <Input
                type="number"
                style={inputStyleWide}
                value={loanYears}
                onChange={handleLoanYearsChange}
                data-testid="loan-years"
              />
            </div>
          </div>
        </div>
      </div>
      <div className="disclaimerWrap">
        <span className="disclaimerText">Please read the disclaimer &gt;</span>
      </div>
    </div>
  );
}

export default MortgageCalculator;
