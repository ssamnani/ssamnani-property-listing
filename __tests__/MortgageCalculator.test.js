import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import MortgageCalculator from "../components/MortgageCalculator";
import { render, fireEvent } from "@testing-library/react";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render("<MortgageCalculator />", div);
});

// Snapshot Testing
it("matches snapshot", () => {
  const tree = renderer.create(<MortgageCalculator />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("should set initial states and render info correctly", () => {
  const { getByText, getByTestId } = render(<MortgageCalculator />);
  expect(getByTestId("monthly-amount")).toHaveTextContent("RM 1886");
  expect(getByTestId("loan-amount")).toHaveTextContent("450000");
  expect(getByTestId("property-price")).toHaveValue(500000);
  expect(getByTestId("down-payment")).toHaveValue(10);
  expect(getByTestId("interest-rate")).toHaveValue(3.6);
  expect(getByTestId("loan-years")).toHaveValue(35);
  expect(getByText("%")).toHaveClass("selected");
});

it("should calculate correct loan amount and monthly payment when input values change", () => {
  const { getByTestId } = render(<MortgageCalculator />);
  const inpPropertyPrice = getByTestId("property-price");
  const inpDownPayment = getByTestId("down-payment");
  const inpInterestRate = getByTestId("interest-rate");
  const inpLoanYears = getByTestId("loan-years");
  const spanMonthlyAmount = getByTestId("monthly-amount");
  const spanLoanAmount = getByTestId("loan-amount");

  fireEvent.change(inpPropertyPrice, { target: { value: 370000 } });
  fireEvent.change(inpDownPayment, { target: { value: 20 } });
  fireEvent.change(inpInterestRate, { target: { value: 4.2 } });
  fireEvent.change(inpLoanYears, { target: { value: 15 } });

  expect(spanMonthlyAmount).toHaveTextContent("RM 2219");
  expect(spanLoanAmount).toHaveTextContent("Loan Amount: RM 296000");
});

it("should render down-payment metric with % as selected by default", () => {
  const { getByText } = render(<MortgageCalculator />);
  const btnPercent = getByText("%");

  expect(btnPercent).toHaveClass("selected");
});

it("should convert down-payment to currency and percentage when clicked", () => {
  const { getByText } = render(<MortgageCalculator />);
  const btnCurrency = getByText("RM");
  const btnPercent = getByText("%");

  fireEvent.click(btnCurrency);
  expect(btnCurrency).toHaveClass("selected");

  fireEvent.click(btnPercent);
  expect(btnPercent).toHaveClass("selected");
});

it("should not change selected toggle when alread selected button is clicked", () => {
  const { getByText } = render(<MortgageCalculator />);
  const btnCurrency = getByText("RM");
  const btnPercent = getByText("%");

  // Click on already selected
  fireEvent.click(btnPercent);

  // Expected no change in selected classes
  expect(btnPercent).toHaveClass("selected");
  expect(btnCurrency).not.toHaveClass("selected");
});
