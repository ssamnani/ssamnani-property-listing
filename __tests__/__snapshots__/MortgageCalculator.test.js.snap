// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="moCalculator"
>
  <div
    className="mortgageTitleBar"
  >
    <span
      className="morgCalcTitle"
    >
      Mortgage Calculator
    </span>
    <span
      className="companyLogoWrap"
    >
      <img
        src="./assets/images/iproperty-logo-black.png"
      />
    </span>
  </div>
  <div
    className="moCalcBody"
  >
    <div
      className="moCalcLeft"
    >
      <div
        className="monthlyAmount"
      >
        <span
          data-testid="monthly-amount"
        >
          RM 
          1886
        </span>
        <span
          className="pm"
        >
          p/m
        </span>
      </div>
      <div
        className="loanAmount"
      >
        <span
          data-testid="loan-amount"
        >
          Loan Amount: RM 
          450000
        </span>
      </div>
    </div>
    <div
      className="moCalcRight"
    >
      <div
        className="moInputWrap"
      >
        <div>
          <div
            className="moInpLabel"
          >
            Property Price (RM)
          </div>
          <input
            className="ant-input"
            data-testid="property-price"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            style={
              Object {
                "backgroundColor": "rgb(243, 244, 246)",
                "width": 200,
              }
            }
            type="number"
            value={500000}
          />
        </div>
        <div>
          <div
            className="moInpLabel"
          >
            Down Payment
          </div>
          <div
            className="downPaymentInputsWrapper"
          >
            <input
              className="ant-input"
              data-testid="down-payment"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              style={
                Object {
                  "backgroundColor": "rgb(243, 244, 246)",
                  "minWidth": "100px",
                  "width": 120,
                }
              }
              type="number"
              value={10}
            />
            <div
              className="downPaymentToggle"
            >
              <button
                className=""
                onClick={[Function]}
              >
                RM
              </button>
              <button
                className="selected"
                onClick={[Function]}
              >
                %
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="moInputWrap"
      >
        <div>
          <div
            className="moInpLabel"
          >
            Interest Rate (%)
          </div>
          <input
            className="ant-input"
            data-testid="interest-rate"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            style={
              Object {
                "backgroundColor": "rgb(243, 244, 246)",
                "width": 200,
              }
            }
            type="number"
            value={3.6}
          />
        </div>
        <div>
          <div
            className="moInpLabel"
          >
            Loan term (Years)
          </div>
          <input
            className="ant-input"
            data-testid="loan-years"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            style={
              Object {
                "backgroundColor": "rgb(243, 244, 246)",
                "width": 200,
              }
            }
            type="number"
            value={35}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="disclaimerWrap"
  >
    <span
      className="disclaimerText"
    >
      Please read the disclaimer &gt;
    </span>
  </div>
</div>
`;
